### PR TITLE
[SPARK-31100][SQL] Check namespace existens when setting namespace

### PR DESCRIPTION
**Pull request by jackylee-ch**
## What changes were proposed in this pull request?
Check the namespace existence while calling "use namespace", and throw NoSuchNamespaceException if namespace not exists.

### Why are the changes needed?
Users need to know that the namespace does not exist when they try to set a wrong namespace.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Run all suites and add a test for this

### CODE REVIEW

**Comment by stczwd** [comment]
cc @HyukjinKwon @cloud-fan @dongjoon-hyun

**Comment by stczwd** [comment]
cc @rxin @xuanyuanking

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala [hunk 1]
```diff
@@ -106,13 +106,14 @@ class CatalogManager(
...
-    if (currentCatalog.name() == SESSION_CATALOG_NAME) {
-      if (namespace.length != 1) {
+    currentCatalog match {
+      case _ if currentCatalog.name() == SESSION_CATALOG_NAME && namespace.length == 1 =>
+        v1SessionCatalog.setCurrentDatabase(namespace.head)
+      case catalog: SupportsNamespaces if catalog.namespaceExists(namespace) =>
+        logInfo(s"set current namespace to ${namespace.mkString(".")}")
+        _currentNamespace = Some(namespace)
+      case _ =>
         throw new NoSuchNamespaceException(namespace)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala]: It's very likely that the underlying catalog can't report namespace existence directly, but can throw namespace not found error when calling APIs like `loadTable(ident)`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala]: And we should allow users to set current namespace when working with these catalogs.

Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala]: > It's very likely that the underlying catalog can't report namespace existence directly, but can throw namespace not found error when calling APIs like `loadTable(ident)`

First, the user needs to sense whether the namespace exists.
Secondly, Those who cannot obtain the namespace status can return true directly in namespaceExists.
Finally, for those don't support namespaces, users should be allowed to set the defaultNamespace, I will add this.

**Comment by cloud-fan** [comment]
how about this: we check namespace existence if the catalog implements `SupportsNamespace`. Otherwise, we just allow it.

**Comment by stczwd** [comment]
> how about this: we check namespace existence if the catalog implements `SupportsNamespace`. Otherwise, we just allow it.

Ah, that's a better idea.
 thanks @cloud-fan

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/CatalogManagerSuite.scala [hunk 2]
```diff
@@ -108,6 +111,17 @@ class CatalogManagerSuite extends SparkFunSuite {
     assert(v1SessionCatalog.getCurrentDatabase == "default")
     catalogManager.setCurrentNamespace(Array("test2"))
     assert(v1SessionCatalog.getCurrentDatabase == "default")
+
+    // Check namespace existence if currentCatalog implements SupportsNamespaces.
+    conf.setConfString("spark.sql.catalog.dummyNamespace", classOf[DummyNameSpaceCatalog].getName)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/CatalogManagerSuite.scala:116]: We can reuse `InMemoryTableCatalog` and create some namespaces before testing `setCurrentNamespace`

**Review by cloud-fan — APPROVED** [review]
LGTM. This behavior is also consistent with the v1 session catalog.

**Comment by cloud-fan** [comment]
cc @rdblue

**Comment by rdblue** [comment]
Sounds good to me. +1

**Comment by stczwd** [comment]
@cloud-fan @rdblue can we reset build?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 3]
```diff
@@ -1394,6 +1394,7 @@ class DataSourceV2SQLSuite
 
   test("Use: v2 catalog is used and namespace does not exist") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1395]: we should update this test with catalogs that don't implement `SupportsNamespace`

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 4]
```diff
@@ -1418,6 +1419,7 @@ class DataSourceV2SQLSuite
 
     sql("USE testcat")
     testShowCurrentNamespace("testcat", "")
+    sql("create namespace testcat.ns1.ns2")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1422]: please upper case the SQL keywords

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 5]
```diff
@@ -2257,6 +2259,7 @@ class DataSourceV2SQLSuite
     spark.conf.unset(V2_SESSION_CATALOG_IMPLEMENTATION.key)
     val sessionCatalogName = CatalogManager.SESSION_CATALOG_NAME
 
+    sql("create namespace testcat.ns1.ns2")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2262]: ditto

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**